Bugfix/bamboo special char replacement

ungaralex opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [X] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [X] I added screenshots/screencast of my UI changes
- [X] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This should fix the problems described in [ARTEMIS-406](https://jirabruegge.in.tum.de/browse/ARTEMIS-406?filter=-1). If a new programming exercise had special character in its name, we couldn't create a build plan in Bamboo, because we were using the full exercise name in the Bamboo project name.

### Description
<!-- Describe your changes in detail -->
As described in the ticket, I am removing every "illegal" character from the Bamboo project name before creating the build plans.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to ArTEMiS
2. Navigate to Course Administration
3. Click on "Exercise" for one of the courses and create a new programming exercise.
4. Use any exercise name and include any of the following characters: "[, ], {, }, <, >, :, @, /, &, %, \, !, |, #, $, *, ; or ~"
5. ~Tick "Publish Build Plan" at the bottom of the page to create the Bamboo plans.~
6. The exercise should be created successfully and the project name in Bamboo shouldn't contain any of the special characters.

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![bamboo_special_chars](https://user-images.githubusercontent.com/27979674/61596300-c6b1da00-ac01-11e9-83a9-2921ec62c989.gif)

krusche commented:
Publish build Plan should not be necessary and is a special option to allow students to access the Build Plan

ungaralex commented:
Thanks, I removed the step from the test description

ungaralex commented:
@krusche There are now some other problems caused by using special characters in the exercise names, e.g. the name is also used in the pom.xml for the artifactId, so the project can't build. 
Should we just generally forbid special characters and show an error in the front-end? I could also just replace them where necessary, but then we have inconsistencies regarding the exercise/project name in different parts of the application.
On top of that, just allowing alphanumeric characters might be too restrictive, since having names containing underscores, round brackets and dashes makes sense in some cases

krusche commented:
please forbid them in the client **and** replace them directly in the Resource (in case there might be a problem in the client or someone uses a direct REST call) before the programming exercise is saved and before the name is used for generating the Bitbucket, Bamboo, Maven and Eclipse files. I guess for a name, alphanumeric values are fine and we don't need any special characters. 

We could think about allowing `-`, `_`, `(` and `)`, but we would need to make 100% sure that there are no problems with these 4 characters in Bitbucket, Bamboo, pom.xml and Java Eclipse projects. So please check the documentation and try if these 4 characters could be used, then we could think about allowing them as well.

By the way: we need to forbid special characters also for the course short name and the exercise short name. I guess we are already doing that, but please double check that these input values can also not be formatted wrong by the user in the client and are checked for consistency in the server.

Actually another option instead of replacing illegal characters in the server would be to decline the request (using some kind of invalid format error code) and show an error message. Then the user would not be surprised by the result.

ungaralex commented:
- [X] I checked the documentations. So, dashes underscores and spaces are fine, `(` and `)` are not. I allowed those three special characters and tested everything on the test server.
- [X] I also added some alert boxes which pop up under the field the user is currently editing if he entered any illegal String. Those alerts should now be visible for every editable text on the exercise page:
![Uploading bamboo_special_chars.gif…]()
- [X] In the back-end, there were already checks for the shortname, etc. I added a check for the title and return a bad request response with an error header in case of any illegal characters

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -205,6 +206,12 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
+        // CHeck if the exercise title matches regex
+        Matcher titleMatcher = TITLE_NAME_PATTERN.matcher(programmingExercise.getTitle());
+        if (!titleMatcher.matches()) {
+            return ResponseEntity.badRequest().headers(HeaderUtil.createAlert(applicationName, "The title is invalid", "titleInvalid")).body(null);
```
This is not terrible, but it would be better to return RequestUtils.badRequest(), catch the error in the client and then display a translatable error.

krusche commented on src/main/webapp/i18n/en/exercise.json:
```diff
@@ -66,6 +66,21 @@
+                    "pattern": "The package name must have at least 1 dot and must not start with a number"
+                },
+                "maxScore": {
+                    "pattern": "The maximum score hast to be an integer between 1 and 9999!"
```
```suggestion
                    "pattern": "The maximum score must be between 1 and 9999!"
```
